### ISSUE TITLE
dotnet publish doesn't copy references with privateassets=all to publish directory

### ISSUE THREAD

**Issue by rohit21agrawal** [dotnet/cli#8340]
Repro steps:

1) Create a new netcoreapp or netstandard library and add net46 as framework as well
2) add a package dependency like newtonsoft.json and set privateassets=all
3) run dotnet publish --self-contained for both frameworks

newtonsoft.json is copied to publish dir for net46 while it is not copied to publish dir for netstandard/netcoreapp frameworks.

CC: @livarcocc @rrelyea

**Comment by rohit21agrawal** [comment]
@wli3 this is a bug that is blocking you from excluding certain references from the output nuspec file in the dotnet tool package.

**Comment by livarcocc** [comment]
I believe the intent of private assets is that it should not be published at all and that it is just a build time dependency.

**Comment by rohit21agrawal** [comment]
we discussed with @emgarten and team, we believe that should not be the case. PrivateAssets=all dictate how the assets would flow to the parent project, and does not indicate a build time dependency, although it can be used like that sometimes.

**Comment by emgarten** [comment]
> From the point of publish, "ExcludeAssets=runtime" or the lack of "runtimes" in IncludeAssets is what should dictate whether it's a build time dependency or not.

That is correct, unless I'm missing something this should all be handled by the assets file which contains a separate compile (build) and runtime (publish) sections for this reason. 

PrivateAssets is a way of setting `ExcludeAssets` for projects that reference you, but it should have zero impact on the project itself.

**Comment by emgarten** [comment]
Originally project.json had a property on packages called `type=build` before NuGet handled this. Later when NuGet added the concept of include, exclude, and suppressParent in project.json it converted type=build into suppressParent=all. 

It may have been the case that the build was handling this flag also because DNU didn't originally support this. 

With `IncludeAssets`, `ExcludeAssets`, and `PrivateAssets` the original `type=build` would become `ExcludeAssets=runtime` and `PrivateAssets=all`.  

`ExcludeAssets=runtime` means that publish should ignore the package, and NuGet will clear that section of the target library.

`PrivateAssets=all` means that the package does not appear in parent projects.

tl;dr: DNU and NuGet 3.2.0 and lower needed the build to handle type=build, but this is no longer the case, and there are 2 knobs for controlling type=build instead of 1 now.

**Comment by eerhardt** [comment]
See https://github.com/NuGet/Home/issues/3507

We had these discussions in the .NET Core SDK v1 timeframe, and `PrivateAssets=All` was the decision at the time.  I can dig up the emails, if necessary.

If we need to revisit this design, let's do it wholistically.

**Comment by emgarten** [comment]
The issue was logged after the NuGet team took a look at packing dotnet tools. There a subset of the dependencies need to be suppressed, but still copied to the publish output folder. We can find other ways to remove dependency entries from the nuspec for pack, but it wouldn't be as clean.

**Comment by dasMulli** [comment]
The ugly workaround at the moment is to do
```xml
<CopyLocalLockFileAssemblies>true</CopyLocalLockFileAssemblies>
```

which will include the assemblies in the publish directory and suppress adding the dependency to the nuspec.
Used e.g. [here](https://github.com/dotnet/sdk/issues/1846#issuecomment-356726385)

**Comment by rohit21agrawal** [comment]
@wli3 how did you work around this in the dotnet tool implementation? did you use the workaround suggested by @dasMulli  ?

**Comment by wli3** [comment]
@rohit21agrawal there is another flag can override private asset. I set all private asset and then set all publish to true, so they won't show up in package reference but in publish folder.

```xml
<Publish>true</Publish>
```

But I am not sure this is the solution for this specific thread

**Comment by livarcocc** [comment]
Btw, I am going to close this issue. Not sure we are ready to re-open this particular can of worms.

**Comment by rohit21agrawal** [comment]
I am fine with closing it as the workaround by @wli3 should work to unblock this scenario for everyone.
